Incorrect parsing of cache age/jitter in keystore

When initializing a keystore with JWKs, keystore.go tries to fetch `age` from `Cache-Control` header.
Unfortunately, when the server responds with something like 
```
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
```
the value extracted from *getCacheAge(cacheControl string)*  is equals to 0 (zero).
Subsequently, `jitter` is set to 0.
This causes an error in subsequent functions, which expect a positive integer as value (see https://golang.org/pkg/math/rand/#Rand.Int63n): 
```
n := rand.Int63n(int64(ks.jitter))

```

A simple fix could be setting a lower bound for jitter in `getCacheJitter` (for example defaultCacheJitter/12).
 

Stacktrace
```
panic: invalid argument to Int63n

goroutine 1 [running]:
math/rand.(*Rand).Int63n(0xc000092150, 0x0, 0xc00019de30)
	/snap/go/3947/src/math/rand/rand.go:111 +0x11e
math/rand.Int63n(...)
	/snap/go/3947/src/math/rand/rand.go:319
github.com/smallstep/certificates/authority/provisioner.(*keyStore).nextReloadDuration(0xc00019de30, 0x0, 0xed4a2c25b)
	/home/mat/go/src/github.com/smallstep/certificates/authority/provisioner/keystore.go:88 +0x3c
github.com/smallstep/certificates/authority/provisioner.newKeyStore(0xc000230100, 0x1d, 0xc000230100, 0x1d, 0x0)
	/home/mat/go/src/github.com/smallstep/certificates/authority/provisioner/keystore.go:46 +0x20f
github.com/smallstep/certificates/authority/provisioner.(*OIDC).Init(0xc0001c8240, 0xc0002222a0, 0xc0002222a8, 0xc0002222b0, 0xc00022229d, 0xc00016da10, 0x3, 0x3, 0xc00016da40, 0x3, ...)
	/home/mat/go/src/github.com/smallstep/certificates/authority/provisioner/oidc.go:146 +0x28e
github.com/smallstep/certificates/authority.(*AuthConfig).Validate(0xc00016d830, 0xc00016da10, 0x3, 0x3, 0xc00016da40, 0x3, 0x3, 0x0, 0xc0001273f8)
	/home/mat/go/src/github.com/smallstep/certificates/authority/config.go:84 +0x275
github.com/smallstep/certificates/authority.(*Config).Validate(0xc000194d00, 0x1b, 0xc00016aa48)
	/home/mat/go/src/github.com/smallstep/certificates/authority/config.go:167 +0x302
github.com/smallstep/certificates/authority.New(0xc000194d00, 0x0, 0x0, 0x0, 0x30, 0x28, 0xad4240)
	/home/mat/go/src/github.com/smallstep/certificates/authority/authority.go:45 +0x40
github.com/smallstep/certificates/ca.(*CA).Init(0xc00016d9e0, 0xc000194d00, 0x2, 0x2, 0x9c6a3d)
	/home/mat/go/src/github.com/smallstep/certificates/ca/ca.go:89 +0x8b
github.com/smallstep/certificates/ca.New(0xc000194d00, 0xc0001276b8, 0x2, 0x2, 0x0, 0xb29eed, 0x4)
	/home/mat/go/src/github.com/smallstep/certificates/ca/ca.go:75 +0xaf
main.startAction(0xc0001d8b00, 0xc000190300, 0xc0001d8b00)
	/home/mat/go/src/github.com/smallstep/certificates/cmd/step-ca/main.go:190 +0x1e4
github.com/smallstep/certificates/vendor/github.com/urfave/cli.HandleAction(0xa43e60, 0xb5e230, 0xc0001d8b00, 0x0, 0xc00016a9c0)
	/home/mat/go/src/github.com/smallstep/certificates/vendor/github.com/urfave/cli/app.go:501 +0xc8
github.com/smallstep/certificates/vendor/github.com/urfave/cli.Command.Run(0xb2acbf, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/mat/go/src/github.com/smallstep/certificates/vendor/github.com/urfave/cli/command.go:165 +0x487
main.main.func4(0xc0001d8840, 0x0, 0x0)
	/home/mat/go/src/github.com/smallstep/certificates/cmd/step-ca/main.go:152 +0x1ae
github.com/smallstep/certificates/vendor/github.com/urfave/cli.HandleAction(0xa43e60, 0xc00016efc0, 0xc0001d8840, 0x0, 0x0)
	/home/mat/go/src/github.com/smallstep/certificates/vendor/github.com/urfave/cli/app.go:501 +0xc8
github.com/smallstep/certificates/vendor/github.com/urfave/cli.(*App).Run(0xc000214000, 0xc00000c080, 0x2, 0x2, 0x0, 0x0)
	/home/mat/go/src/github.com/smallstep/certificates/vendor/github.com/urfave/cli/app.go:268 +0x5aa
main.main()
	/home/mat/go/src/github.com/smallstep/certificates/cmd/step-ca/main.go:155 +0x640
```

--- a/authority/provisioner/keystore.go
+++ b/authority/provisioner/keystore.go
@@ -138,12 +138,10 @@ func getExpirationTime(age time.Duration) time.Time {
 	return time.Now().Truncate(time.Second).Add(age)
 }
 
-// abs returns the absolute value of n using the two's complement form.
-//
-// It will overflow with math.MinInt64 in the same way a branching version
-// would, this is not a problem because maxAgeRegex will block negative numbers
-// and the logic will never produce that number.
+// abs returns the absolute value of n.
 func abs(n time.Duration) time.Duration {
-	y := n >> 63
-	return (n ^ y) - y
+	if n < 0 {
+		return -n
+	}
+	return n
 }

--- a/authority/provisioner/keystore.go
+++ b/authority/provisioner/keystore.go
@@ -18,7 +18,7 @@ const (
 	defaultCacheJitter = 1 * time.Hour
 )
 
-var maxAgeRegex = regexp.MustCompile("max-age=([0-9]*)")
+var maxAgeRegex = regexp.MustCompile("max-age=([0-9]+)")
 
 type keyStore struct {
 	sync.RWMutex
@@ -81,13 +81,13 @@ func (ks *keyStore) reload() {
 	ks.Unlock()
 }
 
+// nextReloadDuration would return the duration for the next rotation. If age is
+// 0 it will randomly rotate between 0-12 hours, but every time we call to Get
+// it will automatically rotate.
 func (ks *keyStore) nextReloadDuration(age time.Duration) time.Duration {
 	n := rand.Int63n(int64(ks.jitter))
 	age -= time.Duration(n)
-	if age < 0 {
-		age = 0
-	}
-	return age
+	return abs(age)
 }
 
 func getKeysFromJWKsURI(uri string) (jose.JSONWebKeySet, time.Duration, error) {
@@ -125,6 +125,10 @@ func getCacheJitter(age time.Duration) time.Duration {
 	switch {
 	case age > time.Hour:
 		return defaultCacheJitter
+	case age == 0:
+		// Avoids a 0 jitter. The duration is not important as it will rotate
+		// automatically on each Get request.
+		return defaultCacheJitter
 	default:
 		return age / 3
 	}
@@ -132,4 +136,14 @@ func getCacheJitter(age time.Duration) time.Duration {
 
 func getExpirationTime(age time.Duration) time.Time {
 	return time.Now().Truncate(time.Second).Add(age)
+}
+
+// abs returns the absolute value of n using the two's complement form.
+//
+// It will overflow with math.MinInt64 in the same way a branching version
+// would, this is not a problem because maxAgeRegex will block negative numbers
+// and the logic will never produce that number.
+func abs(n time.Duration) time.Duration {
+	y := n >> 63
+	return (n ^ y) - y
 }

--- a/authority/provisioner/keystore_test.go
+++ b/authority/provisioner/keystore_test.go
@@ -91,6 +91,49 @@ func Test_keyStore(t *testing.T) {
 	assert.True(t, hits.Hits > 1, fmt.Sprintf("invalid number of hits: %d is not greater than 1", hits.Hits))
 }
 
+func Test_keyStore_noCache(t *testing.T) {
+	srv := generateJWKServer(2)
+	defer srv.Close()
+
+	ks, err := newKeyStore(srv.URL + "/no-cache")
+	assert.FatalError(t, err)
+	defer ks.Close()
+	ks.RLock()
+	keySet1 := ks.keySet
+	ks.RUnlock()
+	// The keys will rotate on Get.
+	// So we won't be able to find the cached ones
+	assert.Len(t, 2, keySet1.Keys)
+	assert.Len(t, 0, ks.Get(keySet1.Keys[0].KeyID))
+	assert.Len(t, 0, ks.Get(keySet1.Keys[1].KeyID))
+	assert.Len(t, 0, ks.Get("foobar"))
+
+	ks.RLock()
+	keySet2 := ks.keySet
+	ks.RUnlock()
+	if reflect.DeepEqual(keySet1, keySet2) {
+		t.Error("keyStore did not rotated")
+	}
+
+	// The keys will rotate on Get.
+	// So we won't be able to find the cached ones
+	assert.Len(t, 2, keySet2.Keys)
+	assert.Len(t, 0, ks.Get(keySet2.Keys[0].KeyID))
+	assert.Len(t, 0, ks.Get(keySet2.Keys[1].KeyID))
+	assert.Len(t, 0, ks.Get("foobar"))
+
+	// Check hits
+	resp, err := srv.Client().Get(srv.URL + "/hits")
+	assert.FatalError(t, err)
+	hits := struct {
+		Hits int `json:"hits"`
+	}{}
+	defer resp.Body.Close()
+	err = json.NewDecoder(resp.Body).Decode(&hits)
+	assert.FatalError(t, err)
+	assert.True(t, hits.Hits > 1, fmt.Sprintf("invalid number of hits: %d is not greater than 1", hits.Hits))
+}
+
 func Test_keyStore_Get(t *testing.T) {
 	srv := generateJWKServer(2)
 	defer srv.Close()
@@ -115,6 +158,34 @@ func Test_keyStore_Get(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			if gotKeys := tt.ks.Get(tt.args.kid); !reflect.DeepEqual(gotKeys, tt.wantKeys) {
 				t.Errorf("keyStore.Get() = %v, want %v", gotKeys, tt.wantKeys)
+			}
+		})
+	}
+}
+
+func Test_abs(t *testing.T) {
+	maxInt64 := time.Duration(1<<63 - 1)
+	minInt64 := time.Duration(-1 << 63)
+	type args struct {
+		n time.Duration
+	}
+	tests := []struct {
+		name string
+		args args
+		want time.Duration
+	}{
+		{"ok", args{0}, 0},
+		{"ok", args{-time.Hour}, time.Hour},
+		{"ok", args{time.Hour}, time.Hour},
+		{"ok maxInt64", args{maxInt64}, maxInt64},
+		{"ok minInt64 + 1", args{minInt64 + 1}, maxInt64},
+		{"overflow on minInt64", args{minInt64}, minInt64},
+		{"overflow on minInt64", args{minInt64}, -minInt64},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := abs(tt.args.n); got != tt.want {
+				t.Errorf("abs() = %v, want %v", got, tt.want)
 			}
 		})
 	}

--- a/authority/provisioner/utils_test.go
+++ b/authority/provisioner/utils_test.go
@@ -658,7 +658,7 @@ func generateJWKServer(n int) *httptest.Server {
 		return ret
 	}
 
-	defaultKeySet := must(generateJSONWebKeySet(2))[0].(jose.JSONWebKeySet)
+	defaultKeySet := must(generateJSONWebKeySet(n))[0].(jose.JSONWebKeySet)
 	srv := httptest.NewUnstartedServer(nil)
 	srv.Config.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		hits.Hits++
@@ -670,8 +670,12 @@ func generateJWKServer(n int) *httptest.Server {
 		case "/openid-configuration", "/.well-known/openid-configuration":
 			writeJSON(w, openIDConfiguration{Issuer: "the-issuer", JWKSetURI: srv.URL + "/jwks_uri"})
 		case "/random":
-			keySet := must(generateJSONWebKeySet(2))[0].(jose.JSONWebKeySet)
+			keySet := must(generateJSONWebKeySet(n))[0].(jose.JSONWebKeySet)
 			w.Header().Add("Cache-Control", "max-age=5")
+			writeJSON(w, getPublic(keySet))
+		case "/no-cache":
+			keySet := must(generateJSONWebKeySet(n))[0].(jose.JSONWebKeySet)
+			w.Header().Add("Cache-Control", "no-cache, no-store, max-age=0, must-revalidate")
 			writeJSON(w, getPublic(keySet))
 		case "/private":
 			writeJSON(w, defaultKeySet)

